src/lib/cache/subCache.cpp

```diff
@@ -0,0 +1,1105 @@
+    {
+      return cSubP;
+    }
+
+    cSubP = cSubP->next;
+  }
+
+  return NULL;
+}
+
+
+
+/* ****************************************************************************
+*
+* subCacheUpdateStatisticsIncrement - 
+*/
+void subCacheUpdateStatisticsIncrement(void)
+{
+  ++subCache.noOfUpdates;
+}
+
+
+
+/* ****************************************************************************
+*
+* subCacheItemInsert - 
+*/
+void subCacheItemInsert(CachedSubscription* cSubP)
+{
+  //
+  // First of all, insertion is done at the end of the list, so, 
+  // cSubP->next must ALWAYS be zero at insertion time
+  //
+  cSubP->next = NULL;
+
+  LM_T(LmtSubCache, ("inserting sub '%s', lastNotificationTime: %lu", cSubP->subscriptionId, cSubP->lastNotificationTime));
+
+  ++subCache.noOfInserts;
+
+  // First insertion?
+  if ((subCache.head == NULL) && (subCache.tail == NULL))
+  {
+    subCache.head   = cSubP;
+    subCache.tail   = cSubP;
+
+    return;
+  }
+
+  subCache.tail->next  = cSubP;
+  subCache.tail        = cSubP;
+}
+
+
+
+/* ****************************************************************************
+*
+* subCacheItemInsert - create a new sub, fill it in, and add it to cache
+*/
+void subCacheItemInsert
+(
+  const char*               tenant,
+  const char*               servicePath,
+  SubscribeContextRequest*  scrP,
+  const char*               subscriptionId,
+  int64_t                   expirationTime,
+  int64_t                   throttling,
+  Format                    notifyFormat,
+  bool                      notificationDone,
+  int64_t                   lastNotificationTime
+)
+{
+  //
+  // Add the subscription to the subscription cache.
+  // But only if any of the entities in entityIdVector is pattern based -
+  // AND it is a subscription of type ONCHANGE
+  //
+
+  //
+  // 00. Check that the subscription is pattern based an ONCHANGE
+  //
+  bool onchange = false;
+
+  for (unsigned int ix = 0; ix < scrP->notifyConditionVector.size(); ++ix)
+  {
+    if (strcasecmp(scrP->notifyConditionVector[ix]->type.c_str(), ON_CHANGE_CONDITION) == 0)
+    {
+      onchange = true;
+      break;
+    }
+  }
+
+  if (!onchange)
+  {
+    return;
+  }
+
+    
+  //
+  // Now allocate the subscription
+  //
+  CachedSubscription* cSubP = new CachedSubscription();
+  LM_T(LmtSubCache,  ("allocated CachedSubscription at %p", cSubP));
+
+
+  //
+  // 1. First the non-complex values
+  //
+  cSubP->tenant                = (tenant[0] == 0)? NULL : strdup(tenant);
+  cSubP->servicePath           = strdup(servicePath);
+  cSubP->subscriptionId        = strdup(subscriptionId);
+  cSubP->reference             = strdup(scrP->reference.get().c_str());
+  cSubP->expirationTime        = expirationTime;
+  cSubP->throttling            = throttling;
+  cSubP->lastNotificationTime  = lastNotificationTime;
+  cSubP->notifyFormat          = notifyFormat;
+  cSubP->next                  = NULL;
+  cSubP->count                 = (notificationDone == true)? 1 : 0;
```

Comment:
This is a new item in the cache (after a subscription-creation). count is 1 or 0, depending on whether it provoked a notification or not